replace deprecated g_type_class_add_private

c-holtermann opened the pull request:
g_type_class_add_private() is deprecated as described in https://gitlab.gnome.org/GNOME/glib/commit/7e5db31d36532274c2b2428ef9bace796928785e

This PR replaces its use with variations of G_DEFINE_TYPE.

There are two cpp and multiple c files affected.

For the cpp files: I'm not experienced with c/c++ but this seems to work. Please have a look at it.
For the c files: I'm trying to get a grip on them as well so I'm working on this PR. The work is not yet done.

jralls commented on libgnucash/engine/qofinstance.cpp:
```diff
@@ -114,7 +114,7 @@ typedef struct QofInstancePrivate
    (G_TYPE_INSTANCE_GET_PRIVATE ((o), QOF_TYPE_INSTANCE,  QofInstancePrivate))
 
-QOF_GOBJECT_GET_TYPE(QofInstance, qof_instance, G_TYPE_OBJECT, {});
+QOF_GOBJECT_GET_TYPE(QofInstance, qof_instance, G_TYPE_OBJECT, G_ADD_PRIVATE(QofInstance));
```
I don't think that ```QOF_GOBJECT_FOO``` really bring much to the party and they're used in only 8 places so let's just use ```G_DEFINE_TYPE_WITH_PRIVATE``` here and in qofbook.cpp instead. Otherwise this is fine, proceed to the rest.

c-holtermann commented on libgnucash/engine/qofbook.cpp:
```diff
@@ -107,7 +107,7 @@ qof_book_option_num_autoreadonly_changed_cb (GObject *gobject,
 #define PARAM_NAME_NUM_AUTOREAD_ONLY "autoreadonly-days"
 
-QOF_GOBJECT_GET_TYPE(QofBook, qof_book, QOF_TYPE_INSTANCE, {});
+G_DEFINE_TYPE_WITH_PRIVATE(QofBook, qof_book, QOF_TYPE_INSTANCE);
```
Is there private data ? Otherwise G_DEFINE_TYPE_WITH_CODE would be sufficient.

c-holtermann replied:
Doesn't even compile.

c-holtermann replied:
G_DEFINE_TYPE_WITH_CODE is the one here

jralls commented on libgnucash/engine/qofbook.cpp:
```diff
@@ -107,7 +107,7 @@ qof_book_option_num_autoreadonly_changed_cb (GObject *gobject,
 #define PARAM_NAME_NUM_AUTOREAD_ONLY "autoreadonly-days"
 
-QOF_GOBJECT_GET_TYPE(QofBook, qof_book, QOF_TYPE_INSTANCE, {});
+G_DEFINE_TYPE_WITH_CODE(QofBook, qof_book, QOF_TYPE_INSTANCE, {});
```
If there's no code shouldn't it be ```G_DEFINE_TYPE```?

c-holtermann replied:
good question

c-holtermann replied:
Compiles fine, makes more sense. I'll use that.

c-holtermann commented:
gnucash/gnome-utils/gnc-main-window.c is more of a challenge.

c-holtermann commented on gnucash/import-export/qif-imp/gnc-plugin-qif-import.c:
```diff
@@ -59,36 +61,9 @@ typedef struct GncPluginQifImportPrivate
-
-    return gnc_plugin_qif_import_type;
-}
-
```
https://github.com/GNOME/glib/blob/6528887534bc42698eebabe33a2c4603aac3dc46/gobject/gtype.h#L1967
looks quite similar, but to change these files seems to be a dive into gobject depths...

c-holtermann replied:
gnucash doesn't start with this change. Gets stuck in loading. I revert it until I understand why.

c-holtermann replied:
fixed -  wrong parent class in G_DEFINE_TYPE_WITH_PRIVATE (as John pointed out below)

c-holtermann commented on gnucash/gnome/search-owner.c:
```diff
@@ -73,33 +73,7 @@ enum
-    }
-
-    return type;
-}
```
This compiles and gnucash runs but I'm not really sure which dialogue this is about. So I haven't tested that yet.

jralls replied:
It's the find customer/vendor/employee dialog. Internally those are all subclasses of owner.

You know that you can run ```gcc -E``` on a file to see what actual C code macros emit, right?

c-holtermann replied:
No, I didn't. I was just trying to find out how I could get that output. Thanks!

c-holtermann commented on gnucash/gnome/gnc-plugin-page-invoice.c:
```diff
@@ -292,34 +292,6 @@ static GObjectClass *parent_class = NULL;
-
-    return gnc_plugin_page_invoice_type;
-}
-
```
Same thing here. Compiles fine but gets stuck in loading. Need to compare the get_type provided by macro with the one in the source. Or maybe the DEFINE_WITH_PRIVATE arguments are wrong, specifically the type.

c-holtermann replied:
I wonder if there is a macro that doesn't create a get_type function so that the original ones could stay.

jralls replied:
Umm, this one doesn't appear to need private info so it should just be ```G_DEFINE_TYPE```... but I don't see ```G_DEFINE_TYPE (GncPluginPageInvoice, gnc_plugin_page_invoice, GNC_TYPE_PLUGIN_PAGE)``` in the commit. Not defining it will certainly cause the plugin to not load.

c-holtermann replied:
It has g_type_class_add_private(klass, sizeof(GncPluginPageInvoicePrivate)); so there is private info. I've been rewriting history by force pushing here.

c-holtermann replied:
Not loading happens when I remove the existing gnc_plugin_page_invoice_get_type in line 295-321 and use a macro of the G_DEFINE_TYPE family. In this case G_DEFINE_TYPE_WITH_PRIVATE in line 325.
https://github.com/Gnucash/gnucash/pull/439/files/d6f9ccacd67892a54b381a911b833d01f1b5ed97..ea499a5b712870b0c5fd6defcb22443f1de42aac#diff-8bd91e5a24f28088b263673ae92e2946R325

jralls commented on gnucash/gnome/gnc-plugin-page-invoice.c:
```diff
@@ -350,6 +322,8 @@ gnc_plugin_page_invoice_new (InvoiceWindow *iw)
 }
 
+G_DEFINE_TYPE_WITH_PRIVATE(GncPluginPageInvoice, gnc_plugin_page_invoice, GNC_TYPE_PLUGIN_PAGE_INVOICE)
+
```
```G_DEFINE_TYPE_WITH_PRIVATE(GncPluginPageInvoice, gnc_plugin_page_invoice, GNC_TYPE_PLUGIN_PAGE_INVOICE)```
There's the problem: You're declaring the class as its own parent. The line should be ```G_DEFINE_TYPE_WITH_PRIVATE(GncPluginPageInvoice, gnc_plugin_page_invoice, GNC_TYPE_PLUGIN_PAGE)```

c-holtermann replied:
Yes, I see. You are right. I thought that was the class itself. Need to look through the other changes about that as well. 

https://developer.gnome.org/gobject/stable/gobject-Type-Information.html#G-DEFINE-TYPE-WITH-PRIVATE:CAPS

the last parameter is T_P | The GType of the parent type

c-holtermann replied:
works ! compiles and loads.

c-holtermann replied:
also did the trick fpr qif import

c-holtermann commented:
I've gone through most of the files. About 14 missing yet. The abovementioned issues need to be fixed (reverted commits).

c-holtermann commented:
I overwrote 934afff969bcb99ce3c231fc7e51aa7915d2e438. The associated conversation had gone. It's still there. I wonder if it stays or will be deleted by github.

c-holtermann commented:
wanted to be sure that this stays:
```
#70 "/usr/local/src/gnucash/gnucash-git-fork-c.holtermann/gnucash/gnome-utils/gnc-tree-view-price.c" 3 4
((void *) 0)
#70 "/usr/local/src/gnucash/gnucash-git-fork-c.holtermann/gnucash/gnome-utils/gnc-tree-view-price.c"
  ;
static gint GncTreeViewPrice_private_offset;

static void
gnc_tree_view_price_class_intern_init (gpointer klass)
{
  gnc_tree_view_price_parent_class = g_type_class_peek_parent (klass);
  if (GncTreeViewPrice_private_offset != 0)
    g_type_class_adjust_private_offset (klass,
					&GncTreeViewPrice_private_offset);
  gnc_tree_view_price_class_init ((GncTreeViewPriceClass *) klass);
}

__attribute__ ((__unused__))
     static inline gpointer
       gnc_tree_view_price_get_instance_private (GncTreeViewPrice * self)
{
  return (((gpointer)
	   ((guint8 *) (self) + (glong) (GncTreeViewPrice_private_offset))));
}

GType
gnc_tree_view_price_get_type (void)
{
  static volatile gsize g_define_type_id__volatile = 0;
  if ((__extension__ (
		       {
		       typedef char
		       _GStaticAssertCompileTimeAssertion_1[(sizeof
							     *
							     (&g_define_type_id__volatile)
							     ==
							     sizeof
							     (gpointer)) ? 1 :
							    -1]
		       __attribute__ ((__unused__));
		       (void) (0 ? (gpointer) *
			       (&g_define_type_id__volatile) : 0);
		       (!(__extension__ (
						     {
						     typedef char
						     _GStaticAssertCompileTimeAssertion_2
						     [(sizeof
						       *
						       (&g_define_type_id__volatile)
						       ==
						       sizeof (gpointer)) ? 1
						      : -1]
						     __attribute__ ((__unused__)); guint64 gapg_temp = __atomic_load_8 ((&g_define_type_id__volatile), 5); (gpointer) gapg_temp;}
			  ))
			&& g_once_init_enter (&g_define_type_id__volatile));}
       )))
    {
      GType g_define_type_id = gnc_tree_view_price_get_type_once ();
      (__extension__ (
		       {
		       typedef char
		       _GStaticAssertCompileTimeAssertion_3[(sizeof
							     *
							     (&g_define_type_id__volatile)
							     ==
							     sizeof
							     (gpointer)) ? 1 :
							    -1]
		       __attribute__ ((__unused__));
		       (void) (0 ? *(&g_define_type_id__volatile) =
			       (g_define_type_id) : 0);
		       g_once_init_leave ((&g_define_type_id__volatile),
					  (gsize) (g_define_type_id));}
       ));
    }
  return g_define_type_id__volatile;
}

__attribute__ ((noinline))
     static GType gnc_tree_view_price_get_type_once (void)
{
  GType g_define_type_id =
    g_type_register_static_simple ((gnc_tree_view_get_type ()),
				   g_intern_static_string
				   ("GncTreeViewPrice"),
				   sizeof (GncTreeViewPriceClass),
				   (GClassInitFunc) (void (*)(void))
				   gnc_tree_view_price_class_intern_init,
				   sizeof (GncTreeViewPrice),
				   (GInstanceInitFunc) (void (*)(void))
				   gnc_tree_view_price_init,
				   (GTypeFlags) 0);
  {
    {
      {
	GncTreeViewPrice_private_offset =
	  g_type_add_instance_private (g_define_type_id,
				       sizeof (GncTreeViewPricePrivate));
      };
    }
  }
  return g_define_type_id;
}
```
from discussion on lost commit 934afff969bcb99ce3c231fc7e51aa7915d2e438.

c-holtermann commented:
Another overwritten commit with discussion: 87f35090756da4d87390e59f88afcfbdb197aba9.

jralls commented:
They're not exactly lost commits, you just rebased them out of the branch history and force-pushed the rebased branch. The commits are still in the repository for now, though they'll eventually get garbage collected because there's no ref pointing to them. 

Is everything working now?

c-holtermann commented:
Quite meets my definition of lost. Or halfway lost. Until the (scary) garbage collector (man) fetches them.

There's still some work to do. The files mentioned in https://github.com/Gnucash/gnucash/pull/439#issuecomment-445188620 are still to be dealt with.

c-holtermann commented:
The comment works as a strike out list for me. I remove the files as I fix them as well as the corresponding reverts. In the end this could be put together as one big commit. What I don't like about this approach and is an obvious drawback of rewriting history like that is the insecurity of the discussion comments.

c-holtermann commented:
I understood that only after commits from force pushed files seemingly vanished. I have now copied into the main conversation what I thought should be saved.

c-holtermann commented:
I'm just now working on the one kind of issue where G_IMPLEMENT_INTERFACE needs to be added. That is maybe five files. Then there are some of the files that have further issues. I don't think that I'll get through it it this weekend as I have some other stuff to attend. Maybe until next weekend it's possible to get all files through.

c-holtermann commented:
Next issue to approach is that of the conflicting types. Reverted commit 6bdea490a3e839eb4dd4894a42dd1f69d8688437 of file gnc-embedded-window.c leads to this error.

This is the output.

```
                 from /usr/local/src/gnucash/gnucash-git-fork-c.holtermann/gnucash/gnome-utils/gnc-embedded-window.c:28:
/usr/local/src/gnucash/gnucash-git-fork-c.holtermann/gnucash/gnome-utils/gnc-embedded-window.c:161:44: error: conflicting types for ‘gnc_embedded_window_init’
 G_DEFINE_TYPE_WITH_CODE(GncEmbeddedWindow, gnc_embedded_window, GTK_TYPE_BOX,
                                            ^~~~~~~~~~~~~~~~~~~
/usr/include/glib-2.0/gobject/gtype.h:1951:17: note: in definition of macro ‘_G_DEFINE_TYPE_EXTENDED_BEGIN_PRE’
 static void     type_name##_init              (TypeName        *self); \
                 ^~~~~~~~~
/usr/include/glib-2.0/gobject/gtype.h:1612:56: note: in expansion of macro ‘_G_DEFINE_TYPE_EXTENDED_BEGIN’
 #define G_DEFINE_TYPE_WITH_CODE(TN, t_n, T_P, _C_)     _G_DEFINE_TYPE_EXTENDED_BEGIN (TN, t_n, T_P, 0) {_C_;} _G_DEFINE_TYPE_EXTENDED_END()
                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/src/gnucash/gnucash-git-fork-c.holtermann/gnucash/gnome-utils/gnc-embedded-window.c:161:1: note: in expansion of macro ‘G_DEFINE_TYPE_WITH_CODE’
 G_DEFINE_TYPE_WITH_CODE(GncEmbeddedWindow, gnc_embedded_window, GTK_TYPE_BOX,
 ^~~~~~~~~~~~~~~~~~~~~~~
/usr/local/src/gnucash/gnucash-git-fork-c.holtermann/gnucash/gnome-utils/gnc-embedded-window.c:53:13: note: previous declaration of ‘gnc_embedded_window_init’ was here
 static void gnc_embedded_window_init (GncEmbeddedWindow *window, GncEmbeddedWindowClass *klass);
             ^~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/src/gnucash/gnucash-git-fork-c.holtermann/gnucash/gnome-utils/gnc-embedded-window.c:197:1: error: conflicting types for ‘gnc_embedded_window_init’
 gnc_embedded_window_init (GncEmbeddedWindow *window,
 ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/glib-2.0/gobject/gobject.h:24,
                 from /usr/include/glib-2.0/gobject/gbinding.h:29,
                 from /usr/include/glib-2.0/glib-object.h:23,
                 from /usr/include/glib-2.0/gio/gioenums.h:28,
                 from /usr/include/glib-2.0/gio/giotypes.h:28,
                 from /usr/include/glib-2.0/gio/gio.h:26,
                 from /usr/include/gtk-3.0/gdk/gdkapplaunchcontext.h:28,
                 from /usr/include/gtk-3.0/gdk/gdk.h:32,
                 from /usr/include/gtk-3.0/gtk/gtk.h:30,
                 from /usr/local/src/gnucash/gnucash-git-fork-c.holtermann/gnucash/gnome-utils/gnc-embedded-window.c:28:
/usr/local/src/gnucash/gnucash-git-fork-c.holtermann/gnucash/gnome-utils/gnc-embedded-window.c:161:44: note: previous declaration of ‘gnc_embedded_window_init’ was here
 G_DEFINE_TYPE_WITH_CODE(GncEmbeddedWindow, gnc_embedded_window, GTK_TYPE_BOX,
                                            ^~~~~~~~~~~~~~~~~~~
/usr/include/glib-2.0/gobject/gtype.h:1951:17: note: in definition of macro ‘_G_DEFINE_TYPE_EXTENDED_BEGIN_PRE’
 static void     type_name##_init              (TypeName        *self); \
                 ^~~~~~~~~
/usr/include/glib-2.0/gobject/gtype.h:1612:56: note: in expansion of macro ‘_G_DEFINE_TYPE_EXTENDED_BEGIN’
 #define G_DEFINE_TYPE_WITH_CODE(TN, t_n, T_P, _C_)     _G_DEFINE_TYPE_EXTENDED_BEGIN (TN, t_n, T_P, 0) {_C_;} _G_DEFINE_TYPE_EXTENDED_END()
                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/src/gnucash/gnucash-git-fork-c.holtermann/gnucash/gnome-utils/gnc-embedded-window.c:161:1: note: in expansion of macro ‘G_DEFINE_TYPE_WITH_CODE’
 G_DEFINE_TYPE_WITH_CODE(GncEmbeddedWindow, gnc_embedded_window, GTK_TYPE_BOX,
 ^~~~~~~~~~~~~~~~~~~~~~~
/usr/local/src/gnucash/gnucash-git-fork-c.holtermann/gnucash/gnome-utils/gnc-embedded-window.c:161:44: error: ‘gnc_embedded_window_init’ used but never defined [-Werror]
 G_DEFINE_TYPE_WITH_CODE(GncEmbeddedWindow, gnc_embedded_window, GTK_TYPE_BOX,
                                            ^~~~~~~~~~~~~~~~~~~
/usr/include/glib-2.0/gobject/gtype.h:1951:17: note: in definition of macro ‘_G_DEFINE_TYPE_EXTENDED_BEGIN_PRE’
 static void     type_name##_init              (TypeName        *self); \
                 ^~~~~~~~~
/usr/include/glib-2.0/gobject/gtype.h:1612:56: note: in expansion of macro ‘_G_DEFINE_TYPE_EXTENDED_BEGIN’
 #define G_DEFINE_TYPE_WITH_CODE(TN, t_n, T_P, _C_)     _G_DEFINE_TYPE_EXTENDED_BEGIN (TN, t_n, T_P, 0) {_C_;} _G_DEFINE_TYPE_EXTENDED_END()
                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/src/gnucash/gnucash-git-fork-c.holtermann/gnucash/gnome-utils/gnc-embedded-window.c:161:1: note: in expansion of macro ‘G_DEFINE_TYPE_WITH_CODE’
 G_DEFINE_TYPE_WITH_CODE(GncEmbeddedWindow, gnc_embedded_window, GTK_TYPE_BOX,

```

Same issue for some more of the remaining files.

c-holtermann commented:
One other thing with the commodities. With this fix I get a message when opening the commodity window:
sys:1: Warning: g_type_class_add_private() called multiple times for the same type
another one is probably unrelated:
(gnucash:24327): Gtk-WARNING **: 02:15:51.194: Negative content width -12 (allocation 1, extents 6x7) while allocating gadget (node button, owner GtkButton)

jralls commented:
I don't see any changes to gnc-embedded-window.c in the branch. Did you revert and rebase them away?

c-holtermann commented:
@jralls - I reverted the changes. The modifications are in https://github.com/Gnucash/gnucash/commit/6bdea490a3e839eb4dd4894a42dd1f69d8688437. Gnucash doesn't compile with the changes.

jralls commented:
Right, because you got the interface decl scrambled. You're declaring and defining the interface in gnc-embedded-window.[ch]. In gnc-embedded-window.h you want 
```C
#define GNC_EMBEDDED_WINDOW gnc_embedded_window_get_type () \
G_DECLARE_INTERFACE (GncEmbeddedWindow, gnc_embedded_window, GNC, EMBEDDED_WINDOW, GObject)
```
and in gnc-embedded-window.c
```C
G_DEFINE_INTERFACE (GncEmbeddedWindow, gnc_embedded_window, G_TYPE_OBJECT)

static void
gnc_embedded_window_default_init (GncEmbeddedWindow *iface)
{
...
}
```
```G_IMPLEMENT_INTERFACE``` goes in ```G_DEFINE_TYPE_WITH_CODE``` for SxEditorDialog in dialog-sx-editor.c and dialog-sx-editor2.c.

c-holtermann commented:
The problem with the gnc_embedded_window_default_init is, that there is an added class parameter. As far as I see it is only being used for memory checking purposes. I removed it in my recent patch. Seems to work but ```
gnc_gobject_tracking_remember(G_OBJECT(window),
                                  G_OBJECT_CLASS(klass));
``` is missing. I don't know if that's a problem.

c-holtermann commented on gnucash/gnome-utils/gnc-embedded-window.c:
```diff
@@ -194,20 +193,20 @@ gnc_embedded_window_class_init (GncEmbeddedWindowClass *klass)
     gtk_orientable_set_orientation (GTK_ORIENTABLE(window), GTK_ORIENTATION_VERTICAL);
 
     // Set the style context for this widget so it can be easily manipulated with css
-    gnc_widget_set_style_context (GTK_WIDGET(window), "GncEmbededWindow");
```
Is this in use ? `grep embeded * -Ri` doesn't lead to a result. I think it's save to correct the spelling.

jralls commented:
Re: The GncEmbeddedWindow style context (Github won't let me reply directly, dunno why) setting css styling is something a user might do, but it won't work with the misspelled class name so that change is better than safe, it's a fix to a bug nobody's encountered yet.

jralls commented:
Re: The GncEmbeddedWindow style context (Github won't let me reply directly, dunno why) setting css styling is something a user might do, but it won't work with the misspelled class name so that change is better than safe, it's a fix to a bug nobody's encountered yet.

Re: gnc_gobject_tracking_remember: It's not used for this class so it's safe to remove, just be sure to remove the corresponding gnc_gobject_tracking_forget as well.

c-holtermann commented:
As far as I see it should even be possible to keep the memory tracking as it is possible to have a NULL pointer as klass. 
```
if (klass == NULL)
        klass = G_OBJECT_GET_CLASS(object);
```
that is `gnc_gobject_tracking_remember (GObject *object, GObjectClass *klass)` in gnucash/gnome-utils/gnc-gobject-utils.h.

c-holtermann commented:
Using gnc_gobject_tracking_remember with NULL as class parameter seems to work. Fixed these files:

gnucash/gnome-utils/gnc-plugin.c 3ba5e9ca21037cc87b726aa52cc675c34fcf7f51 - fix in fab25bd508c127f1b5e4b0ec0d6012ec87142a89
gnucash/gnome-utils/gnc-plugin-page.c 96d8d6d1a1c0d3c1bc348fc4c00906f88de13907 - fix in a407f5c3609d424fc30360a9c4f249cf40c2a994
gnucash/gnome-utils/gnc-embedded-window.c 1f2288bbaae4146e0043f39adbae68e2501bda23 - fix in https://github.com/Gnucash/gnucash/pull/439/commits/04744ef3328536d6fd25f93bb7daf31a41733dad and f101bf0671fbe58ca89c0527cc128901cf5b9e8b
gnucash/gnome-utils/gnc-tree-model.c 0daeb412f043e5b3beaad74dcfe9eeb43fa9ccc5 - fix in db8574005572c4dd5c8d2e0bb041de8a4f3e0a93
gnucash/gnome-utils/gnc-tree-view.c 2b845fb21cbdb48e379344a28d5f14e6bbaaf0b1 - fix in c99bccd1fde80a286840339126a1734c96c3c3f2

problem in 
gnucash/gnome-utils/gnc-main-window.c 3db9965f80eb796e6e5b1444d8776af7bbca9004 - partial fix in f9da955ff4a40bf61e31721ebe0954f8ec89a6df
as mention above. Reverted. Needs looking into.

unhandled and yet to do:
gnucash/gnome-utils/gnc-combott.c 23dc9adb906d0f71a425e01b9d1949eddc98f6aa

Quite some progress. A little remains to do. Reviews desired.

c-holtermann commented:
Thank you John. I don't know in which way these tracking things are being used. I just wanted to be able to keep them. If they are not needed then away with them.
I think this whole thing can be finished this year. So from my side there is no hurry. After all issues are solved I'd like to clean up the commit history to make it rather logic than chronologic. Maybe cherry-picking to a 2nd pull request and drop this one. If the commits remain in regular branch log but the pull request is dropped, will the conversations remain ? I like to be able to access development history.

jralls commented:
Yes, if you close this PR and make a new one from your rebase that will preserve the comments in this one. Cleaning up history to a logical progression (I like one commit per module for large changes like this, please make sure that the build works and all tests pass for each commit) is a good plan.

I'd like to get this merged in time for releasing 3.4 on the 30th. Do you think you can manage that?

c-holtermann commented:
Yes, I think getting things done until 30th should be possible. I hope to have worked through all the remaining issues so that the code works until the end of the weekend. That leaves enough time for testing, reviewing, rebasing.

c-holtermann commented:
main-window: invalid cast fixed but Segmentation fault still remains.

c-holtermann commented:
gdb stack trace:
```
#0  0x00007ffff7f67951 in gnc_plugin_page_account_tree_init () at /home/christoph/.local/lib64/libgnc-gnome.so
#1  0x00007ffff6f7eb2a in g_type_create_instance () at /usr/lib64/libgobject-2.0.so.0
#2  0x00007ffff6f61688 in  () at /usr/lib64/libgobject-2.0.so.0
#3  0x00007ffff6f62d85 in g_object_new_with_properties () at /usr/lib64/libgobject-2.0.so.0
#4  0x00007ffff6f63941 in g_object_new () at /usr/lib64/libgobject-2.0.so.0
#5  0x00007ffff7f67550 in gnc_plugin_page_account_tree_new () at /home/christoph/.local/lib64/libgnc-gnome.so
#6  0x00007ffff7f69096 in gnc_plugin_page_account_tree_recreate_page () at /home/christoph/.local/lib64/libgnc-gnome.so
#7  0x00007ffff30be269 in gnc_plugin_page_recreate_page () at /home/christoph/.local/lib64/gnucash/libgncmod-gnome-utils.so
#8  0x00007ffff30ac4ae in gnc_main_window_restore_page () at /home/christoph/.local/lib64/gnucash/libgncmod-gnome-utils.so
#9  0x00007ffff30ad10d in gnc_main_window_restore_window () at /home/christoph/.local/lib64/gnucash/libgncmod-gnome-utils.so
#10 0x00007ffff30ad4ab in gnc_main_window_restore_all_windows () at /home/christoph/.local/lib64/gnucash/libgncmod-gnome-utils.so
#11 0x00007ffff7fa3303 in gnc_restore_all_state () at /home/christoph/.local/lib64/libgnc-gnome.so
#12 0x00007ffff27cf7e4 in call_c_hook () at /home/christoph/.local/lib64/gnucash/libgncmod-engine.so
#13 0x00007ffff7b61084 in g_hook_list_marshal () at /usr/lib64/libglib-2.0.so.0
#14 0x00007ffff27cf958 in gnc_hook_run () at /home/christoph/.local/lib64/gnucash/libgncmod-engine.so
#15 0x00007ffff30a2fdd in gnc_book_opened () at /home/christoph/.local/lib64/gnucash/libgncmod-gnome-utils.so
#16 0x00007ffff30a3e80 in gnc_post_file_open () at /home/christoph/.local/lib64/gnucash/libgncmod-gnome-utils.so
#17 0x00007ffff30a4065 in gnc_file_open_file () at /home/christoph/.local/lib64/gnucash/libgncmod-gnome-utils.so
#18 0x0000000000406df5 in inner_main ()
#19 0x00007ffff7cc973d in  () at /usr/lib64/libguile-2.0.so.22
#20 0x00007ffff7ca1bca in  () at /usr/lib64/libguile-2.0.so.22
#21 0x00007ffff7d3ce4a in  () at /usr/lib64/libguile-2.0.so.22
#22 0x00007ffff7caba23 in scm_call_4 () at /usr/lib64/libguile-2.0.so.22
#23 0x00007ffff7ca2331 in  () at /usr/lib64/libguile-2.0.so.22
#24 0x00007ffff7ca2415 in scm_c_with_continuation_barrier () at /usr/lib64/libguile-2.0.so.22
#25 0x00007ffff7d23c96 in  () at /usr/lib64/libguile-2.0.so.22
#26 0x00007ffff107f3b5 in GC_call_with_stack_base () at /usr/lib64/libgc.so.1
#27 0x00007ffff7d24098 in scm_with_guile () at /usr/lib64/libguile-2.0.so.22
#28 0x00007ffff7cc9932 in scm_boot_guile () at /usr/lib64/libguile-2.0.so.22
```

jralls commented on gnucash/gnome/gnc-plugin-page-account-tree.c:
```diff
@@ -446,10 +446,12 @@ gnc_plugin_page_account_tree_init (GncPluginPageAccountTree *plugin_page)
-        g_object_set_data(G_OBJECT(plugin_page), PLUGIN_PAGE_IMMUTABLE,
-                          GINT_TO_POINTER(1));
+    if (page_list != NULL) {
+        if (plugin_page == page_list->data)
```
It's a bit more ideomatic to write that as
```C
  if (page_list && plugin_page == page_list->data)
```
instead of nested ```if```s.

c-holtermann replied:
But does that work ? It's the access to page_list->data that breaks the code. Even in testing it. Aren't all members of this comparison being checked and the problem would still occur ? I'll test it.

gjanssens commented:
Op zaterdag 15 december 2018 12:37:41 CET schreef Christoph Holtermann:
> c-holtermann commented on this pull request.
> 
> > @@ -446,10 +446,12 @@ gnc_plugin_page_account_tree_init
> > (GncPluginPageAccountTree *plugin_page)
>      /* Is this the first accounts page? */
>      page_list =
>          gnc_gobject_tracking_get_list(GNC_PLUGIN_PAGE_ACCOUNT_TREE_NAME);
> -    if (plugin_page == page_list->data)
> -    {
> -        g_object_set_data(G_OBJECT(plugin_page), PLUGIN_PAGE_IMMUTABLE,
> -                          GINT_TO_POINTER(1));
> +    if (page_list != NULL) {
> +        if (plugin_page == page_list->data)
> 
> But does that work ? It's the access to page_list->data that breaks the
> code. Even in testing it. Aren't all members of this comparison being
> checked and the problem would still occur ? I'll test it.

Yes this works because conditions are evaluated left to right. And if the 
first condition is false there's no need to evaluate the second condition in 
the AND test (&&) as it would not change the outcome. So the compiler writes 
code that does skip the second test. This has worked for as long as I know.

c-holtermann commented:
updated progress status:

Using gnc_gobject_tracking_remember with NULL as class parameter seems to work. Fixed these files:

gnucash/gnome-utils/gnc-plugin.c 3ba5e9ca21037cc87b726aa52cc675c34fcf7f51 - fix in fab25bd508c127f1b5e4b0ec0d6012ec87142a89
gnucash/gnome-utils/gnc-plugin-page.c 96d8d6d1a1c0d3c1bc348fc4c00906f88de13907 - fix in a407f5c3609d424fc30360a9c4f249cf40c2a994
gnucash/gnome-utils/gnc-embedded-window.c 1f2288bbaae4146e0043f39adbae68e2501bda23 - fix in https://github.com/Gnucash/gnucash/pull/439/commits/04744ef3328536d6fd25f93bb7daf31a41733dad and f101bf0671fbe58ca89c0527cc128901cf5b9e8b
gnucash/gnome-utils/gnc-tree-model.c 0daeb412f043e5b3beaad74dcfe9eeb43fa9ccc5 - fix in db8574005572c4dd5c8d2e0bb041de8a4f3e0a93
gnucash/gnome-utils/gnc-tree-view.c 2b845fb21cbdb48e379344a28d5f14e6bbaaf0b1 - fix in c99bccd1fde80a286840339126a1734c96c3c3f2
gnucash/gnome-utils/gnc-main-window.c 3db9965f80eb796e6e5b1444d8776af7bbca9004 - partial fix in f9da955ff4a40bf61e31721ebe0954f8ec89a6df and latest: https://github.com/Gnucash/gnucash/pull/439/commits/c877290571b1742909c9f423567f6082bf5943b5

Having a null parameter for klass in gnc_gobject_tracking_remember doesn't seem to work as desired so this needs looking into. Maybe removing at some points as John suggested.

unhandled and yet to do:
gnucash/gnome-utils/gnc-combott.c 23dc9adb906d0f71a425e01b9d1949eddc98f6aa

Quite some progress. A little remains to do. Reviews desired.

gjanssens commented:
> Works.
> 
> Account Tree Tab should be immutable if it is the only one. The condition takes care of conditions when there is more than one. (Does that ever happen ?) In case of null it should be immutable.

It can happen. The user can open more than one Account Tree Tab.

gjanssens commented on gnucash/gnome-utils/gnc-date-format.c:
```diff
@@ -76,7 +76,7 @@ struct _GNCDateFormatPriv
 
 #define GNC_DATE_FORMAT_GET_PRIVATE(o)  \
-   (G_TYPE_INSTANCE_GET_PRIVATE ((o), GNC_TYPE_DATE_FORMAT, GNCDateFormatPriv))
+   (G_TYPE_INSTANCE_GET_PRIVATE ((o), GNC_TYPE_DATE_FORMAT, GNCDateFormatPrivate))
```
[Emmanuel Bassi's article](https://www.bassi.io/articles/2013/06/21/the-king-is-dead/) you refer to on the mailing list thread also suggests replacing calls to MY_OBJECT_GET_PRIVATE with my_object_get_instance_private calls.
While I wouldn't do it as part of this PR it may be a nice follow up cleanup action.

jralls commented on gnucash/gnome-utils/gnc-embedded-window.c:
```diff
@@ -233,20 +193,18 @@ gnc_embedded_window_class_init (GncEmbeddedWindowClass *klass)
-    gnc_gobject_tracking_remember(G_OBJECT(window),
-                                  G_OBJECT_CLASS(klass));
+    gnc_gobject_tracking_remember(G_OBJECT(window), NULL);
     LEAVE(" ");
```
The answer on the G_DEFINE_TYPE bug](https://gitlab.gnome.org/GNOME/glib/issues/1621) was "too bad, just unroll the macro", which suggests an alternative: Copy the  G_DEFINE_TYPE.* macros that we need to GNC_DEFINE_TYPE.* and modify GNC_DEFINE_TYPE to use a correct GInstanceInitFunc.

c-holtermann replied:
That sounds like a possible way to go. I'll try it. Thank's for doing the request to the glib people.

c-holtermann replied:
I'll put correcting the name of the css class into a separate pull request.

c-holtermann replied:
John, would you redefine the macro or define similar ones that provide an init function with two parameters ? And where would you put these definitions. They should surely go to a central location. And should we mix the ones with one parameter and that with two depending on the use case ? I suppose in the current use case it would affect only G_DEFINE_TYPE_WITH_CODE which could be changed to G_DEFINE_TYPE_WITH_CODE_INIT_TWO_PARAMS.

jralls replied:
Neither. You want to create **GNC**_DEFINE_TYPE_WITH_CODE that calls _**GNC**_DEFINE_TYPE_EXTENDED_BEGIN. The latter will be a copy of [_G_DEFINE_TYPE_EXTENDED_BEGIN](https://gitlab.gnome.org/GNOME/glib/blob/master/gobject/gtype.h#L1949) except that the first line will be
```C
static void     type_name##_init              (TypeName        *self, void *class); \
```
so that it will work with classes having two-parameter init functions.
I suppose it should go in gnucash/gnome-utils/gnc-gobject-utils.h.

c-holtermann replied:
ok. Sounds good.

jralls commented:
>It can happen. The user can open more than one Account Tree Tab.

A bit off-topic for this PR but that shouldn't require a different tree model, just another tree view. The only case I can think of that we'd have more than one model of the same class is registers.

c-holtermann commented:
The last commits have removed the last occurences of g_type_class_add_private.

c-holtermann commented:
Next I'll go through the suggestions from the reviews.

c-holtermann commented on gnucash/gnome-utils/gnc-currency-edit.c:
```diff
@@ -185,7 +154,7 @@ gnc_currency_edit_get_property (GObject    *object,
 
-
+G_DEFINE_TYPE_WITH_PRIVATE(GNCCurrencyEdit, gnc_currency_edit, GTK_TYPE_COMBO_BOX)
 
```
Ok, I'm not familiar with a right positioning of this macro. I'll do it for this file and if it looks good proceed to all others.

c-holtermann commented on gnucash/gnome-utils/gnc-gobject-utils.h:
```diff
@@ -110,6 +110,37 @@ void gnc_gobject_tracking_dump (void);
+  _GNC_DEFINE_TYPE_EXTENDED_BEGIN_PRE(TypeName, type_name, TYPE_PARENT) \
+  _G_DEFINE_TYPE_EXTENDED_BEGIN_REGISTER(TypeName, type_name, TYPE_PARENT, flags) \
+
+
```
Redefining this macro means checking different glib versions if they are still compatible. I ran into trouble trying to implement this. I found three different versions in different versions.

jralls commented:
Crud, I pushed too much. Please rebase away 1a5c2dd and 6d533e3 when you squish this down... which I think we're now ready for. Is there anything left that I've forgotten?

c-holtermann commented:
I changed the macro to GNC_DEFINE_TYPE_WITH_CODE in all occurences of init function with two parameters.

John, you wanted me to move the macro up to the definitions. That has to take place in a lot of files. I already did it where I used the GNC_ macro.

Are we still on the road that I put up a 2nd pull request to have a clean history ? I'd prefer that because there is a lot of revert and back-revert.

jralls commented:
>Are we still on the road that I put up a 2nd pull request to have a clean history ? I'd prefer that because there is a lot of revert and back-revert.

You proposed that as a way to not lose as much of the discussion history on this PR. I'm fine with that if you still want to do it. In your local repository you're going to do a giant ```git rebase -i``` to squash the 138 commits down to a digestible number, maybe a dozen or two. It will probably take several passes to get everything in the right place. Whether you push the end result to this branch or a new one for a new PR doesn't matter that much in the long run except that if you make a new one this one will preserve the false starts and rewrites for posterity.

c-holtermann commented:
I'd like to move things to a new PR for a clean commit history and preserving discussion. I will finish that that over christmas or one day later if that's still fitting with your timeframe.

jralls commented:
NP.  I generally do more-or-less-final test builds, tag the release, make the tarballs, and write the release notes on Saturday. That leaves Sunday to do the Mac and Windows builds. My cut-off for commits is comfortably Noon California time on Saturday, but I can hold it back if there's something last minute that needs to go in.

c-holtermann commented:
Branch for clean up and PR to come: https://github.com/c-holtermann/gnucash/tree/PR-G_ADD_PRIVATE-clean

c-holtermann commented:
John, unfortunately I am busy with a faulty harddrive so that I won't make it until tomorrow. I'd like to finish the rebasing without hurry so I'd like to exclude this patch from the current version.

jralls commented:
Bummer.

jralls commented:
@c-holtermann  do you not have time to finish this? Shall I take it over and finish it up?

c-holtermann commented:
I'll have a look tomorrow. If it's too much I for me I'll hand it to you. Thanks for the offer.

c-holtermann commented:
@jralls - if you have the capacity now I'm fine with you taking over.

jralls commented:
I merged the PR-G_ADD_PRIVATE-clean branch. It turned out it was ready to go and I just didn't realize it.

c-holtermann commented:
thanks
